sal/tests/server/unit/test_data.py

```diff
@@ -0,0 +1,437 @@
+                                       'current':2,
+                                       'modified':[5,]}}}])
+def test_data_tree_post_report(server, json_content):
+
+    """
+    Tests that a DataTree does not put a branch report or a leaf report to a
+    path 
+
+    GIVEN
+        A request with a URL path
+        AND a request with JSON content describing a branch report or leaf
+        report
+    WHEN
+        The DataTree posts to the path
+    THEN
+        An InvalidRequest is raised
+    """
+
+    path = 'this/is/the/path'
+
+    _check_raising_request(server, 'post', path, InvalidRequest,
+                           json=json_content)
+
+
+# Post currently has a bug which passes a different path format for the source
+# and target paths (target is prefixed with '/', source isn't), so this is
+# currently an expected fail
+@pytest.mark.xfail(reason='Known bug when copying using DataTree.post')
+@settings(suppress_health_check=[HealthCheck.function_scoped_fixture])
+@given(source_revision=valid_revisions)
+def test_data_tree_copy_object(source_revision, server,
+                               mock_persistence_provider):
+
+    """
+    Tests that a DataTree correctly copies an object from one path to another
+    path using a PersistenceProvider 
+
+    GIVEN
+        A request with a URL path
+        AND a request with a query with a source=path
+        AND a request with a query with source_revision=non-negative int
+    WHEN
+        The DataTree posts to the path
+    THEN
+        The Datatree copies the content from the PersistenceProvider source
+        path to the PersistenceProvider path
+    """
+
+    path = 'this/is/the/path'
+    source_path = 'this/is/another/path'
+    data = {'source':source_path, 'source_revision':source_revision}
+
+    expected = ('', 204)
+    call_args = '/{}'.format(path), '/{}:{}'.format(source_path,
+                                                    source_revision)
+
+    _check_request(server, 'post', path, expected, data=data)
+
+    mock_persistence_provider.copy.assert_called_with(call_args)
+
+
+@settings(suppress_health_check=[HealthCheck.function_scoped_fixture])
+@given(source_revision=invalid_revisions)
+def test_data_tree_copy_invalid_source_revision(source_revision, server,
+                                                mock_persistence_provider):
+
+    """
+    Tests that a DataTree copy is invalid if the source has a negative source
+    revision 
+
+    GIVEN
+        A request with a URL path
+        AND a request with a query with a source=path
+        AND a request with a query with source_revision=negative int
+    WHEN
+        The DataTree posts to the path
+    THEN
+        A BadRequest exception is raised
+    """
+
+    path = 'this/is/the/path'
+    source_path = 'this/is/another/path'
+    data = {'source':source_path, 'source_revision':source_revision}
+    
+    _check_raising_request(server, 'post', path, BadRequest, data=data)
+
```

Comment:
**suggestion:** There are tests for `GET`, `PUT`, and `COPY`, but none for `DELETE`.

Reply:
Hmmm yeah, I appreciate that seems strange - I think it's because there really isn't any logic in the interface to test with `DELETE`.  I'll add a comment to that effect in the module docstring.